### ISSUE TITLE
Implement way to safely interrupt the ongoing Chaos Experiment

### ISSUE THREAD

**Issue by k8s-dev** [litmuschaos/litmus#932]
Current way to abort the ongoing experiment is to remove the {experiment-name} pod directly by force, which may result in adverse effect. In fact- rather than reducing the blast-radius - it might increase it and thus is not a recommended way to handle the abort scenarios.

Proposal is to implement the fail-safe mechanism to abort the chaos-experiments mid-way(the big-red-button!)

A rollback mechanism of the chaos-experiment would be icing on the cake, may be a stretch goal for the time being.

**Comment by ksatchit** [comment]
@k8s-dev , you are right. Forced delete of {experiment-name} pod can have adverse impact, as these are run as **jobs**, which, by default will re-run/reschedule the pod until (entrypoint scripts,in our case playbooks) complete -- which means, the chaos you wanted to prevent/stop, can repeat. 

The workaround today is job removal (`kubectl delete job <>` ). 

Note that, at this point jobs still don't have OwnerReference set against them and therefore will not be reconciled (i.e., recreated) by the chaos-operator.

**Comment by ksatchit** [comment]
@k8s-dev , here are some updates: 

- With the introduction of the new [chaos-runner](https://github.com/litmuschaos/chaos-executor.git), we shall be propagating the OwnerReference of the ChaosEngine to the experiment jobs and its dependent resources (such as pumba jobs or other helper litmus daemonsets) as well (they run as independent resources today, though launched by the chaos-runner which itself has the chaosEngine as the owner). 

- With this, removal of the chaosEngine will immediately remove the experiment jobs. Basically, an automated version of the approach mentioned in the [previous comment](https://github.com/litmuschaos/litmus/issues/932#issuecomment-553777242)

We shall continue to post updates as we make progress!

**Comment by ksatchit** [comment]
Here's the latest on this issue: 

- Propagation of OwnerReference proved to be a bit tricky and resulted in a few undesirable outcomes (not least the fact that some resources would have multiple owners - one set by the native K8s controllers and one set by the engine controller/runner/experiment). So, instead of OwnerReference, the ***chaos lineage***, so to say was passed via standard labels (UID of the chaosengine) to all experiment resources
- The abort was then made possible via a json merge patch to the `.spec.engineState` property of the chaosengine, which ensured in the controller reconcile force removing all the experiment resources bearing the UID label.
- A finalizer set on the ChaosEngine at the time of chaos resource  creation is removed only after all the resources are removed upon abort (or graceful completion of experiments)

This is the first phase of abort. The second phase is being tracked by https://github.com/litmuschaos/litmus/issues/1284 where chaos processes started inside target containers need to be stopped

**Comment by ksatchit** [comment]
This has implemented as described here: https://github.com/litmuschaos/litmus/issues/1284#issuecomment-692491108

With this abort is supported for both node (phase-1) & pod (phase-2) level experiments via a patch operation on the ChaosEngine to set `.spec.engineState: stop`
